applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_nonlinear_strain_energy_response_function.cpp

```diff
@@ -0,0 +1,378 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Mahmoud Sesa, https://github.com/mahmoudsesa
+//
+
+// System includes
+
+// External includes
+#include "utilities/variable_utils.h"
+
+// Project includes
+#include "adjoint_nonlinear_strain_energy_response_function.h"
+
+namespace Kratos 
+{
+    AdjointNonlinearStrainEnergyResponseFunction::AdjointNonlinearStrainEnergyResponseFunction(ModelPart& rModelPart, Parameters ResponseSettings)
+    : AdjointStructuralResponseFunction(rModelPart, ResponseSettings)
+    {
+        // initializing mConditionsRHS 
+        for(auto& cond_i : rModelPart.Conditions())
+        {
+            SizeType number_of_nodes = cond_i.GetGeometry().size();
+            SizeType dimension = cond_i.GetGeometry().WorkingSpaceDimension();
+            SizeType vec_size = number_of_nodes * dimension;
+          
+            mConditionsRHS[cond_i.Id()] = ZeroVector(vec_size);
+
+            // This stores a vector of pointers to the model condition as a member variable because I can not use const condition in 
+            mConditions[cond_i.Id()] = rModelPart.pGetCondition(cond_i.Id());
+
+            // initializes a vector that stores the gradient vectors for the last step
+            mResponseGradient_0[cond_i.Id()] = ZeroVector(vec_size);
+        }
+    }
+
+    AdjointNonlinearStrainEnergyResponseFunction::~AdjointNonlinearStrainEnergyResponseFunction()
+    {
+    }
+
+    double AdjointNonlinearStrainEnergyResponseFunction::CalculateValue(ModelPart& rModelPart)
+    {
+        KRATOS_TRY;
+
+        this->CheckForBodyForces(rModelPart);
+
+        ProcessInfo &r_current_process_info = rModelPart.GetProcessInfo();
```

Comment:
const, then you don't need to check with `Has` 5 lines later

Reply:
I can not use const ProcessInfo in CalculateLocalSystem, that's why I am doing it like that